### DISCUSSION TITLE
[general]  tar_group_by() for very large targets

### DISCUSSION THREAD

**Discussion by stemangiola** [ropensci/targets#1147] [General]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I am using `targets` with the `iteration = "group"` paradigm. One of my target is a dataframe, with a lot of rows, that contain very large `SummarizedExperiment` objects.

```r
# A tibble: 8 × 4
  tissue_harmonised data            number_of_chunks se_md5                          
  <chr>             <named list>               <int> <chr>                           
1 muscle            <SmmrzdEx[,40]>                8 1acbfcd21184334526df33d9d26051e5
2 muscle            <SmmrzdEx[,40]>                8 eb3d44c29627f813bcccc467067b3c8f
3 muscle            <SmmrzdEx[,40]>                8 a5c76f268b5872af5d6c0f814c62a9dc
4 muscle            <SmmrzdEx[,40]>                8 638675ac279a2c7f6593e992db847ba4
5 muscle            <SmmrzdEx[,40]>                8 79e3c1be8ce62c1a5741d4a672871c9a
6 muscle            <SmmrzdEx[,40]>                8 01e96aa628dea2c9f0ca988ba733472e
7 muscle            <SmmrzdEx[,40]>                8 948f232310dd0ccaf6a096bfbefc35bd
8 muscle            <SmmrzdEx[,40]>                8 1096efdf5feaf0c5655ce97fdeb4a90d
```

When I do 

```r
	tarchetypes::tar_group_by(
			sce_df_split_by_gene_grouped,
			sce_df_split_by_gene,
			se_md5, 
			deployment = "main"
		)
```

`targets` write a single file `sce_df_split_by_gene_grouped` that appears to duplicate the content of all `sce_df_split_by_gene` files. Is this correct?

I would imagine that a leaner approach could be taken to add just a column index information that would not require to create a monolithic replicate of all the data.

Also, when I do `tar_target(..., map())` does `targets` reads the whole `sce_df_split_by_gene_grouped`? Or there is a `HDF5` on-disk approach, to avoid using a lot of memory during file reading?

tar_group_by() + tar_target(..., map()) need to read the whole data.frame?

(I am using `format = "qs"`, even after reading the documentation, I am not sure if putting each group in a different file for faster loading might depend on the `format` chosen.)

Thanks!

**Comment by wlandau** [comment]
> I would imagine that a leaner approach could be taken to add just a column index information that would not require to create a monolithic replicate of all the data.

An alternative is to use dplyr::group_by() and tar_group() yourself to make the upstream  sce_df_split_by_gene grouped to begin with.

> Also, when I do tar_target(..., map()) does targets reads the whole sce_df_split_by_gene_grouped? Or there is a HDF5 on-disk approach, to avoid using a lot of memory during file reading?

It has to because targets promised an interface for splitting that can only take place on R objects  in memory. If that does not work for you, you could formulate your problem with a different approach, e.g. branching over a smaller dataset and creating each SummarizedExperiment object within each dynamic branch.

**Comment by stemangiola** [comment]
> An alternative is to use dplyr::group_by() and tar_group() yourself to make the upstream sce_df_split_by_gene grouped to begin with.

But `sce_df_split_by_gene` has `iteration = "group"` , would it work anyway?

```r
		tar_target(
			sce_df_split_by_gene,
			pseudobulk_df_tissue |> 
				aggregate() |> 
				map_quantile_scale_abundance() |> 
				map_keep_abundant() |> 
				se_add_dispersion() |> 
				add_number_of_gene_chunks() |> 
				map_split_se_by_gene(data, number_of_chunks),
			pattern = map(pseudobulk_df_tissue),
			iteration = "group",
			resources = tar_resources(crew = tar_resources_crew("small_slurm_memory"))
		)
```
